mypyc/irbuild/prepare.py

```diff
@@ -555,21 +555,53 @@ def add_setter_declaration(
     ir.method_decls[setter_name] = decl
 
 
+def check_matching_args(init_sig: FuncSignature, new_sig: FuncSignature) -> bool:
+    num_init_args = len(init_sig.args) - init_sig.num_bitmap_args
+    num_new_args = len(new_sig.args) - new_sig.num_bitmap_args
+    if num_init_args != num_new_args:
+        return False
+
+    for idx in range(1, num_init_args):
+        init_arg = init_sig.args[idx]
+        new_arg = new_sig.args[idx]
+        if init_arg.type != new_arg.type:
+            return False
+
+        if init_arg.kind != new_arg.kind:
+            return False
+
+    return True
+
+
 def prepare_init_method(cdef: ClassDef, ir: ClassIR, module_name: str, mapper: Mapper) -> None:
     # Set up a constructor decl
     init_node = cdef.info["__init__"].node
+
+    new_node: SymbolNode | None = None
+    new_typeinfo = cdef.info.get("__new__")
+    if new_typeinfo and new_typeinfo.fullname and not new_typeinfo.fullname.startswith("builtins"):
```

Comment:
Add comment explaining what is happening here.

This name seems a bit confusing. `new_typeinfo` refers to a symbol table node? If that's the case, maybe rename it as `new_symbol` or similar.

The prefix should be `"builtins."` (with a dot at the end), since there can be a user-defined module such as `builtinsx`.